src/Arcus.Messaging.Pumps.ServiceBus.KeyRotation/Extensions/IServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,67 @@
+        /// <param name="maximumUnauthorizedExceptionsBeforeRestart">
+        ///     The fallback when the Azure Key Vault notification doesn't get delivered correctly,
+        ///     how many times should the message pump run into an <see cref="UnauthorizedException"/> before restarting.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        ///     Thrown when the <paramref name="services"/> or the searched for <see cref="AzureServiceBusMessagePump"/> based on the given <paramref name="jobId"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="ArgumentException">Thrown when the <paramref name="subscriptionNamePrefix"/> or <paramref name="serviceBusTopicConnectionStringSecretKey"/> is blank.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown when the <paramref name="maximumUnauthorizedExceptionsBeforeRestart"/> is less then zero.</exception>
+        public static IServiceCollection WithReAuthenticationOnNewSecretVersion(
+            this IServiceCollection services,
+            string jobId,
+            string subscriptionNamePrefix,
+            string serviceBusTopicConnectionStringSecretKey,
+            int maximumUnauthorizedExceptionsBeforeRestart = 5)
+        {
+            Guard.NotNull(services, nameof(services), "Requires a collection of services to add the re-authentication background job");
+            Guard.NotNullOrWhitespace(subscriptionNamePrefix, nameof(subscriptionNamePrefix), "Requires a non-blank subscription name of the Azure Service Bus Topic subscription, to receive Azure Key Vault events");
+            Guard.NotNullOrWhitespace(serviceBusTopicConnectionStringSecretKey, nameof(serviceBusTopicConnectionStringSecretKey), "Requires a non-blank secret key that points to a Azure Service Bus Topic");
+            Guard.NotLessThan(maximumUnauthorizedExceptionsBeforeRestart, 0, nameof(maximumUnauthorizedExceptionsBeforeRestart), "Requires an unauthorized exceptions count that is greater than zero");
+
+            services.AddCloudEventBackgroundJob(subscriptionNamePrefix, serviceBusTopicConnectionStringSecretKey);
+            services.WithServiceBusMessageHandler<ReAuthenticateMessageHandler, CloudEvent>(serviceProvider =>
+            {
+                AzureServiceBusMessagePump messagePump =
+                    serviceProvider.GetServices<IHostedService>()
+                                   .OfType<AzureServiceBusMessagePump>()
+                                   .FirstOrDefault(pump => pump.JobId == jobId);
+
+                Guard.NotNull(messagePump, nameof(messagePump), 
+                    $"Cannot register re-authentication without a '{nameof(AzureServiceBusMessagePump)}' with 'JobId' = '{jobId}'");
+
+                messagePump.Settings.Options.MaximumUnauthorizedExceptionsBeforeRestart = maximumUnauthorizedExceptionsBeforeRestart;
```

Comment:
Should we mutate the message pump here?

Reply:
If we do, what if we call a "reconfigure" method instead of changing the options directly?

Reply:
OK, yes, done that.